refactor: Package manager: enhance nuget protocol support, refactor functionality and prepare for future ui improvements

carlosscastro opened the pull request:
Closes #6033 

## Description

- Adds improved support for NuGet protocol
- Refactors package feed functionality into future testeable units (tests incoming in R13)
- Adds concept of query that can be passed to feed
- Server side set up ready for adding search functionality in the UI: search text bot, pre-release check box and paging. Work needed is just build the UI and pass query object
- Next steps for full cleanup:
  - Only allow pointers to service index in package manager
  - Implement searchbox, prerelease check box and paging in package maanger UI
  - When users add a feed, need to specify npm or nuget to avoid using heuristics to decide the type of feed
A more minimal / tactical fix for the myget issue can be found in this [WIP PR](https://github.com/microsoft/BotFramework-Composer/pull/6034)

## Screenshots

Testing: 
Npm feed
![image](https://user-images.githubusercontent.com/21318528/109432572-ad1d6a80-79c0-11eb-900d-e96120553d8c.png)

NuGet 
![image](https://user-images.githubusercontent.com/21318528/109432612-dc33dc00-79c0-11eb-8d8d-1264650de110.png)

MyGet: automatically filtered with tags: bf-component, prerelease, semver version
![image](https://user-images.githubusercontent.com/21318528/109432628-ed7ce880-79c0-11eb-9f59-ea6184281de1.png)

Local

![image](https://user-images.githubusercontent.com/21318528/109432644-f8d01400-79c0-11eb-9fd3-8d76448f04e1.png)

Quick design for PR readability

![image](https://user-images.githubusercontent.com/21318528/109432843-0508a100-79c2-11eb-8ff6-7de8e407ee32.png)

beyackle commented on extensions/packageManager/src/node/index.ts:
```diff
@@ -247,30 +132,30 @@ export default async (composer: IExtensionRegistration): Promise<void> => {
+          const packages = await feed.getPackages(packageQuery);
+
+          if (Array.isArray(packages)) {
+            combined = combined.concat(packages);
```
I think `combined.push(...packages)` is the more idiomatic way to do this.

carlosscastro replied:
This is the original code that just moved and I'm trying to reduce risk. So, I'll change it as part of this PR if I get unit tests in. Otherwise I'll make this change along with unit tests to be sure I exercise this path, since in the current UI this is always a single element array  in practical terms.

carlosscastro replied:
Added this one after all, thanks for the suggestion!

beyackle commented on extensions/packageManager/src/node/feeds/nuget/nugetFeed.ts:
```diff
@@ -0,0 +1,127 @@
+
+    // Now that we have a service index either from seed data or from querying the NuGet feed,
+    // we get the query endpoint, build query parameters, call feed api and return query results to caller
+    const queryEndpoint = serviceIndex?.resources?.find((resource) => resource['@type'] === 'SearchQueryService');
```
Is it possible for the array of resources to contain multiple search query services? If so, we should use `filter` instead of `find` and then iterate over all of them.

carlosscastro replied:
No, just one as per the spec. Thankfully they clarify that.

beyackle commented on extensions/packageManager/src/node/feeds/nuget/nugetFeed.ts:
```diff
@@ -0,0 +1,127 @@
+      this.composer.log(`Retrieving NuGet packages with computed search url ${nuGetSearchUrl}`);
+      const httpResponse = await axios.get(nuGetSearchUrl);
+
+      if (!httpResponse || !httpResponse.data) {
```
`if (httpResponse?.data != null)` would work here (or `if (httpResponse?.data)` if there's no way this can be tricked by a falsy non-null value).

carlosscastro replied:
So here I want the opposite logical result (enter if something is null) but I get your point of doing a single conditional statement.

beyackle commented on extensions/packageManager/src/node/feeds/nuget/nugetFeed.ts:
```diff
@@ -0,0 +1,127 @@
+    // If a query is provided, it takes precedence. Otherwise, default to the package source one.
+    const effectiveQuery = query ?? this.packageSource?.defaultQuery;
+    let url: string = `${baseUrl}?prerelease=${effectiveQuery?.prerelease ?? true}&semVerLevel=${
+      effectiveQuery?.semVerLevel ?? '2.0.0'
```
Why is 2.0.0 the default here?

carlosscastro replied:
This is the default value we chose based on the docs. I can add a link to the spec here.

beyackle reviewed (changes requested):
Mostly looks good - just a few suggestions.

a-b-r-o-w-n commented on extensions/packageManager/src/node/feeds/nuget/localNuGetFeed.ts:
```diff
@@ -0,0 +1,126 @@
+    // * extract only folders from that list
+    // * pass each one through the getPackageInfo function, which extracts metadata from the package
+    // * return a feed in the form that is used by nuget search API
+    const packages = readdirSync(url, { withFileTypes: true }).filter((f) => f.isDirectory());
```
Why not use the async version here?

carlosscastro replied:
Thanks Andy! Great suggestions. I left this code unchanged to take it in stages, and just changed the structure for now. I'll try this out and depending on how much it changes things andthe risk equation, I'll either include in this iteration or in a separate pr

carlosscastro replied:
Actually looks much better and seems low risk, so I'll test it and push these suggestions shortly.

a-b-r-o-w-n commented on extensions/packageManager/src/node/feeds/nuget/localNuGetFeed.ts:
```diff
@@ -0,0 +1,126 @@
+
+    let versions;
+    try {
+      versions = readdirSync(packageDir, { withFileTypes: true })
```
Use the async version here as well.

a-b-r-o-w-n commented on extensions/packageManager/src/node/feeds/nuget/localNuGetFeed.ts:
```diff
@@ -0,0 +1,126 @@
+    const feed = [];
+    for (const p of packages) {
+      feed.push(await this.getPackageInfo(url, p.name));
+    }
```
This looks like it should be pushed in to an array of promises and then return `Promise.all(...)`.

a-b-r-o-w-n reviewed (changes requested):
Overall this is a great refactor. Can you use the async file apis in the local nuget feed?

carlosscastro commented:
Thanks for the review @a-b-r-o-w-n and @beyackle. Pushed code that addresses your feedback. Happy to keep iterating. Also note that this is not the end game, this is the first step towards the end game, mostly enabling some more scenarios for R12, setting up the foundation for unit testing and upcoming functionality in r13.

benbrown commented on extensions/packageManager/src/node/index.ts:
```diff
@@ -249,30 +132,30 @@ export default async (composer: IExtensionRegistration): Promise<void> => {
-          if (Array.isArray(feed)) {
-            combined = combined.concat(feed);
+          const feed: IFeed = await new FeedFactory(composer).build(packageSource);
+          const packageQuery: IPackageQuery = {
```
this blocks the use of the default query, so it undoes the narrower default query.
this should only be used if a default query is not specified.

carlosscastro replied:
Yep, changed to this logic. Added a todo since in reality we want the provided query to further filter with the package query as basis. However we don't need that right now.

carlosscastro replied:
Pushing a fix shortly.

benbrown replied:
thanks!

beyackle commented on extensions/packageManager/src/node/feeds/nuget/localNuGetFeed.ts:
```diff
@@ -0,0 +1,125 @@
+    // * pass each one through the getPackageInfo function, which extracts metadata from the package
+    // * return a feed in the form that is used by nuget search API
+    const packages = await readdir(url, { withFileTypes: true });
+    const feedPromises: Promise<IPackageDefinition>[] = [];
+    for (const p of packages.filter((f) => f.isDirectory())) {
+      feedPromises.push(this.getPackageInfo(url, p.name));
+    }
```
`const feedPromises = packages.filter((f) => f.isDirectory()).map((f) => this.packageInfo(url, f.name));` should work too.

beyackle reviewed (changes requested):
One minor thing - otherwise looks good.
